codegen restructuring

vtjnash opened the pull request:
This creates the function `jl_create_native` for outputting a `llvm::Module` from a list (`jl_array_t`) of functions (`jl_method_instance_t`) which is independent from the JIT (as needed by https://github.com/JuliaGPU/CUDAnative.jl/issues/122). It's now used by the build system also, so that it'll be tested too. This (mostly) eliminates the `shadow_module`, and does a slightly more aggressive job at ensuring functions are getting precompiled.

There's still some more work to do to make that work even better, but that can be done later. I'm also now printing out errors that happen as a result of codegen bugs, rather than propagating them to the user, which makes a few of the tests (for codegen breakage) are a bit noisier now – those will need to be fixed now.

fixes #26767
fixes #26714

maleadt commented on src/jitlayers.h:
```diff
@@ -174,9 +230,9 @@ class JuliaOJIT {
+    DenseMap<void*, StringRef> ReverseLocalSymbolTable;
 };
 extern JuliaOJIT *jl_ExecutionEngine;
-JL_DLLEXPORT extern LLVMContext jl_LLVMContext;
```
Why unexport this? CUDAnative needs access to the Julia context. Can't get it from the module (`Value::getContext` doesn't seem wrapped in the C API), also because we need it _before_ getting the module (eg. to perform `llvmcall`s with LLVM.jl).

Keno replied:
Cxx needs it as well.

maleadt commented:
Cool! How should I access the call targets from CUDAnative in order to build a module with all functions defined?

vtjnash commented:
> How should I access the call targets from CUDAnative 
> Value::getContext

Feel free to add exports from aotlayers.cpp to expose these with C wrappers. I know `jl_create_native` needs some extra arguments (specifically, cgparams) too.

maleadt commented:
> Feel free to add exports from aotlayers.cpp to expose these with C wrappers. I know `jl_create_native` needs some extra arguments (specifically, cgparams) too.

OK. Could you describe the behavior/indented use of `jl_create_native`, and the fields of `jl_native_code_desc_t` then? Does `jl_create_native` compile all necessary functions (ie. no trampolines), without using a cache? And should I still bother with a `module_setup` hook, or just re-configure the datalayout/triple as `jl_dump_native` does?

vtjnash commented:
> And should I still bother with a module_setup

I'm thinking of moving the creation of the Module into the caller, to sidestep this question entirely.

maleadt commented:
What about my other questions?

Also, I was using `julia_type_to_llvm` in generated functions in combination with LLVM.jl. This function now takes a codegen context, which a) isn't available from C/Julia, and b) doesn't exist yet at the time the LLVM.jl-based generated function runs. I guess that b) will become even worse once there isn't a global LLVM context anymore (which I assume is the direction you're going in).
Any comments on how to do this style of LLVM-based custom codegen now?

vtjnash commented:
> I was using julia_type_to_llvm in generated functions in combination with LLVM.jl. This function now takes a codegen context

That other method is not exported – I didn't change the signature of `julia_type_to_llvm` (yet 😛  – it might gain a LLVMContextRef eventually).

> will become even worse once there isn't a global LLVM context anymore (which I assume is the direction you're going in). Any comments on how to do this style of LLVM-based custom codegen now?

Yes, probably true. But I don't really expect `jl_create_native` will be exactly right for your use case long-term anyways, but will just provides a sample usage. Eventually, I expect you'll just provide accessors from this file to be able to do ccalls to do everything that it can do.

maleadt commented:
> That other method is not exported – I didn't change the signature of julia_type_to_llvm (yet 😛 – it might gain a LLVMContextRef eventually).

Ah OK, I didn't see the second definition. An explicit LLVM context should be fine, assuming we can maintain a global one in CUDAnative.

vtjnash committed:
codegen: restructure, remove recursion

also provides support for using a different code_native format,
as a fallback, later we'll want to make this more configurable

there are now several primary interfaces to native code:
 - codegen: mostly internal, support for translating IR to LLVM
 - jitlayers: manages runtime codegen results and executable memory
 - aotcompile: support for managing external code output
 - disasm: pretty-printer for code objects
 - debuginfo: tracking for unwind info

also removes the global type caches and
move all codegen pass handling to aotcompile.cpp

iblis17 commented:
This FreeBSD CI build already ran more than 9 hrs: https://freebsdci.julialang.org/#/builders/1/builds/9298
I manually killed it.

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25984#issuecomment-388126943) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/395e9da7c466bd35fe9709abcbefcbce79fe63f6/c1a7370_vs_ee4b18f/report.md). cc @ararslan

maleadt commented on src/aotcompile.cpp:
```diff
@@ -0,0 +1,960 @@
+    JL_GC_PUSH1(&src);
+    JL_LOCK(&codegen_lock);
+
+    for (int worlds = 2; worlds > 0; worlds--) {
```
all this is pretty dense, could you add some comments?

maleadt commented on src/jitlayers.h:
```diff
@@ -30,23 +27,58 @@ extern "C" {
+    size_t world = 0;
+    const jl_cgparams_t *params = &jl_default_cgparams;
+    bool cache = false;
+} jl_codegen_params_t;
```
do you intend for this to be reusable across code generators, or just sample usage like jl_create_native?

vtjnash replied:
It can be reusable. It's mostly intended to represent a single `llvm::Module` of output, but we can reset it when you compile the result (empty the `globals` list and reset the module handle) to prepare it for more reuse.

vtjnash commented on src/aotcompile.cpp:
```diff
@@ -0,0 +1,947 @@
+//  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x00
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x10
+    0, "NOT", 0, "YY", "XOR", 0, "AND", 0, 0, 0, "MUL", "SUM", 0, "SUB", 0, "DIV", // 0x20
```
I was confused by some build errors. We have run into trouble with certain symbols in the past (primarily `@`, which means version for some linkers), and thus sometimes strip some of them. Here, I'm deciding just to address the issue more completely.

iblis17 commented:
This build got killed by OS OOM killer on one of my worker. (well, it owns 21G ram
```
pid 87376 (julia), uid 1002, was killed: out of swap space
```

JeffBezanson commented:
I get this trying to precompile Plots:
```
signal (11): Segmentation fault
in expression starting at /home/jeff/.julia/packages/Colors/P2rUx/src/algorithms.jl:64
unknown function (ip: (nil))
jl_apply at /home/jeff/src/julia/src/julia.h:1688 [inlined]
do_apply at /home/jeff/src/julia/src/builtins.c:643
jl_f__apply_pure at /home/jeff/src/julia/src/builtins.c:675
pure_eval_call at ./compiler/abstractinterpretation.jl:635
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:69
abstract_call_known at ./compiler/abstractinterpretation.jl:873
abstract_call at ./compiler/abstractinterpretation.jl:895
abstract_call at ./compiler/abstractinterpretation.jl:880
abstract_eval at ./compiler/abstractinterpretation.jl:974
```

vtjnash commented:
OK, lots of use of `@pure` triggered a rare code path where I forgot to handle the `else` case. Fixed.

maleadt commented on src/aotcompile.cpp:
```diff
@@ -0,0 +1,909 @@
+// removes special characters from the name of GlobalObjects,
+// which might cause them to be treated special by LLVM
+// or the system linker
+// the only non-identifier characters we keep are '%' and '.',
```
Could you use an other character? `%` isn't actually safe, and doesn't get past the NVPTX assembler: https://github.com/llvm/llvm-project/blob/aa324c5441f229bd1ff07407affeab95740328fe/llvm/lib/MC/MCAsmInfo.cpp#L103-L106

vtjnash replied:
OK. Alternates might be `$` or use a digram like `..` or be non-reversible and use no signaling character or use a pattern like `_AAB_`. Thoughts/preference?

maleadt replied:
`$` seems like a good alternative, but I'll have to verify not only LLVM's NVPTX but also `ptxax` is happy with that.

maleadt replied:
New version is nvptx/ptxas-compatible.

vtjnash commented on src/codegen.cpp:
```diff
@@ -143,20 +141,19 @@ extern void _chkstk(void);
 bool imaging_mode = false;
 
+// shared llvm state
+static LLVMContext &jl_LLVMContext = *(new LLVMContext());
```
One of the purpose of this change is to eliminate this state object. (There's a function accessor for it in the aotcompiler file)

maleadt replied:
OK, so I can't be assuming a global state anymore starting with this PR? That'll require some more work on my side them :slightly_smiling_face:

maleadt commented:
A non-global LLVM context (but `jl_get_llvm_context` to get one from the AOT compiler output) won't currently work for the GPU compiler, because we implement generated functions with LLVM.jl by returning a `llvm::Function*` that relies on the Julia context. Do you have a short-term plan for that? If not, maybe we should keep the context (which is still global anyway) exported until we have a solution.

Gnimuc commented on src/codegen.cpp:
```diff
@@ -1382,367 +1048,101 @@ static void jl_setup_module(Module *m, const jl_cgparams_t *params = &jl_default
-
-
-extern "C" JL_DLLEXPORT
-void *jl_get_llvmf_decl(jl_method_instance_t *mi, size_t world, bool getwrapper, const jl_cgparams_t params)
```
Is there an alternative function to use instead of this one? It was used in Cxx.jl: https://github.com/JuliaInterop/Cxx.jl/blob/256485c04c76feee515e83f091ba0d6c8f3286ee/src/cxxstr.jl#L107
